task/control-plane-agent/src/mgs_gimlet/host_phase2.rs

```diff
@@ -0,0 +1,285 @@
+        // in our outgoing net task queue).
+        let port = match current.state {
+            State::NeedToSendFirstMgs(port) => {
+                current.state = State::WaitingForFirstMgs {
+                    port,
+                    deadline: now + DELAY_TRY_OTHER_MGS,
+                };
+                port
+            }
+            State::WaitingForFirstMgs { port, deadline } => {
+                if now < deadline {
+                    return None;
+                }
+                // Timed out waiting for a response from the first MGS we tried;
+                // flip to the other one.
+                let port = match port {
+                    SpPort::One => SpPort::Two,
+                    SpPort::Two => SpPort::One,
+                };
+                current.state = State::WaitingForSecondMgs {
+                    port,
+                    deadline: now + DELAY_RETRY,
+                };
+                port
+            }
+            State::WaitingForSecondMgs { port, deadline } => {
+                if now < deadline {
+                    return None;
+                }
+                // Timed out waiting for a response from the second MGS we
+                // tried; flip back to the first and retry.
+                current.retry_count += 1;
+                if current.retry_count >= MAX_ATTEMPTS {
+                    current.notify_calling_task();
+                    self.current = None;
+                    return None;
+                }
+                let port = match port {
+                    SpPort::One => SpPort::Two,
+                    SpPort::Two => SpPort::One,
+                };
+                current.state = State::WaitingForFirstMgs {
+                    port,
+                    deadline: now + DELAY_TRY_OTHER_MGS,
+                };
+                port
+            }
+            State::Fetched => return None,
+        };
+
+        let message = Message {
+            header: Header {
+                version: gateway_messages::version::V2,
+                message_id,
+            },
+            kind: MessageKind::SpRequest(SpRequest::HostPhase2Data {
+                hash: current.hash,
+                offset: current.offset,
+            }),
+        };
+
+        let n = gateway_messages::serialize(tx_buf, &message).unwrap_lite();
+
+        Some(UdpMetadata {
+            addr: SP_TO_MGS_MULTICAST_ADDR,
+            port: MGS_UDP_PORT,
+            size: n as u32,
+            vid: vlan_id_from_sp_port(port),
+        })
+    }
+
+    pub(crate) fn ingest_incoming_data(
+        &mut self,
+        port: SpPort,
+        hash: [u8; 32],
+        offset: u64,
+        data: &[u8],
+    ) {
+        let current = match self.current.as_mut() {
+            Some(current)
+                if hash == current.hash && offset == current.offset =>
+            {
+                current
+            }
+            _ => return,
+        };
+
+        self.buffer.clear();
```

Comment:
Is it possible for there to be data in the buffer that has yet to be used?

Reply:
It's possible we could get multiple packets from MGS that all contain the same data, but otherwise no. Maybe it would be clearer to do something like:

* if our state is already `Fetched`, just return
* if our state is not `Fetched`, assert that `self.buffer.is_empty()`

? The current implementation will always copy in the latest data from MGS (although it should only be sending exact copies, if we get multiples like this).

Reply:
Made that change in 77e52be88c6ae4563c0dccc98efaea18a2cf9728